api/src/main/java/org/apache/iceberg/io/CheckIterator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+/**
+ * An Iterator of {@link Optional} that checks a predicate for each item based on another Iterator.
+ * If predicate is true, the optional item has content, otherwise the optional is empty.
+ *
+ * @param <T> the type of objects of the dependent Iterator
+ */
+public abstract class CheckIterator<T> implements CloseableIterator<Optional<T>> {
```

Comment:
Do we need this class? Or could it be a method in `CloseableIterable`?

```java
  static <E> CloseableIterable<Optional<E>> asOptions(CloseableIterable<E> iterable, Predicate<E> pred) {
    return CloseableIterable.transform(iterable,
        item -> pred.test(item) ? Optional.ofNullable(item) : Optional.empty());
  }
```

Reply:
I see that this is `CloseableIterator` rather than `CloseableIterable`, but there is also a `transform` method for `CloseableIterator` that might work, right?